### PR TITLE
Enhance streaming ingestion metrics

### PR DESCRIPTION
**Pull request by AmatyaAvadhanula**
### Description

Add a metric for partition-wise kafka/kinesis lag for streaming ingestion.

Emit lag metrics for streaming ingestion when supervisor is not suspended and state is in {RUNNING, IDLE, UNHEALTHY_TASKS, UNHEALTHY_SUPERVISOR}

Document metrics

<hr>

This PR has:

- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] a release note entry in the PR description.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/dev/license.md)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java [hunk 1]
```diff
@@ -4095,6 +4096,17 @@ protected void emitLag()
         }
 
         LagStats lagStats = computeLags(partitionLags);
+        for (Map.Entry<PartitionIdType, Long> entry : partitionLags.entrySet()) {
+          emitter.emit(
+              ServiceMetricEvent.builder()
+                                .setDimension(DruidMetrics.DATASOURCE, dataSource)
+                                .setDimension(DruidMetrics.PARTITION, entry.getKey())
```
[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:4103]: Yes, I'll make the change

File: indexing-service/src/main/java/org/apache/druid/indexing/common/actions/LocalTaskActionClient.java [hunk 2]
```diff
@@ -72,14 +73,15 @@ public <RetType> RetType submit(TaskAction<RetType> taskAction)
...
+    emitTimerMetric("task/action/run/time", System.currentTimeMillis() - performStartTime, taskAction);
     return result;
   }
 
-  private void emitTimerMetric(final String metric, final long time)
+  private void emitTimerMetric(final String metric, final long time, TaskAction taskAction)
   {
     final ServiceMetricEvent.Builder metricBuilder = ServiceMetricEvent.builder();
     IndexTaskUtils.setTaskDimensions(metricBuilder, task);
+    metricBuilder.setDimension(DruidMetrics.TASK_ACTION_TYPE, taskAction.getClass().getName());
```
[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/actions/LocalTaskActionClient.java:84]: Thanks, I wasn't aware of it. I'll revert these changes

[the remark this answers is not in the document]
Comment (gianm) [reply indexing-service/src/main/java/org/apache/druid/indexing/common/actions/LocalTaskActionClient.java:84]: I just raised it now, so it's understandable you weren't aware of it!

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (AmatyaAvadhanula) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:4109]: Done

File: extensions-contrib/prometheus-emitter/src/main/resources/defaultMetrics.json [hunk 3]
```diff
@@ -53,9 +53,10 @@
...
+  "ingest/kafka/lag" : { "dimensions" : ["dataSource", "stream"], "type" : "gauge", "help": "Total lag between the offsets consumed by the Kafka indexing tasks and latest offsets in Kafka brokers across all partitions. Minimum emission period for this metric is a minute."},
+  "ingest/kafka/maxLag" : { "dimensions" : ["dataSource", "stream"], "type" : "gauge", "help": "Max lag between the offsets consumed by the Kafka indexing tasks and latest offsets in Kafka brokers across all partitions. Minimum emission period for this metric is a minute."},
+  "ingest/kafka/avgLag" : { "dimensions" : ["dataSource", "stream"], "type" : "gauge", "help": "Average lag between the offsets consumed by the Kafka indexing tasks and latest offsets in Kafka brokers across all partitions. Minimum emission period for this metric is a minute."},
+  "ingest/kafka/avgLag" : { "dimensions" : ["dataSource", "stream", "partition"], "type" : "gauge", "help": "Partition-wise lag between the offsets consumed by the Kafka indexing tasks and latest offsets in Kafka brokers. Minimum emission period for this metric is a minute."},
```
Comment (gianm) [inline extensions-contrib/prometheus-emitter/src/main/resources/defaultMetrics.json:59]: should be `partitionLag`?

Comment (AmatyaAvadhanula) [reply extensions-contrib/prometheus-emitter/src/main/resources/defaultMetrics.json]: Sorry. fixed

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java [hunk 4]
```diff
@@ -4075,9 +4076,14 @@ protected void emitNoticesQueueSize()
...
-      // don't emit metrics if supervisor is suspended or not in a healthy running state
-      // (lag should still available in status report)
+    SupervisorStateManager.State basicState = stateManager.getSupervisorState().getBasicState();
+    boolean unhealthySupervisorOrTasks = SupervisorStateManager.BasicState.UNHEALTHY_TASKS.equals(basicState)
+                                         || SupervisorStateManager.BasicState.UNHEALTHY_SUPERVISOR.equals(basicState);
+
+    if (spec.isSuspended() || !(stateManager.isSteadyState() || stateManager.isIdle() || unhealthySupervisorOrTasks)) {
+      // Don't emit metrics if the supervisor is suspended. Also,
+      // to emit metrics, the state must be in {healthy steady state, idle or UNHEALTHY_TASKS or UNHEALTHY_SUPERVISOR}
+      // (lag should still be available in the status report)
```
Comment (kfaraz) [inline indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:4086]:
```suggestion
      // Emit metrics only if the supervisor state is RUNNING, IDLE, UNHEALTHY_TASKS or UNHEALTHY_SUPERVISOR
      // (lag should still be available in the status report in other supervisor states)
```

File: indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (kfaraz) [reply indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/SeekableStreamSupervisor.java:4109]: We already add the datasource dimension. Will `stream` add any new information?

**Review by kfaraz — APPROVED** [review]

**Pull request merged**